crux-llvm counterexample codegen

Code generation for counterexamples is currently not working, and is generating compile errors like the following:

```
   results/memtest1/counter-example-14.c:6:43: error: use of undeclared identifier 'BV'
   int8_t const crucible_values_int8_t[] = { BV 0 };
```

Seems like we need better pretty printing for values.

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Errors.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Errors.hs
@@ -142,10 +142,10 @@ explainBB = \case
 
 detailBB :: IsExpr (SymExpr sym) => BadBehavior sym -> Doc
 detailBB = \case
-  BBUndefinedBehavior ub -> UB.ppReg ub
+  BBUndefinedBehavior ub -> UB.ppDetails ub
   BBMemoryError me -> ME.details me
 
 ppBB :: IsExpr (SymExpr sym) => BadBehavior sym -> Doc
 ppBB = \case
-  BBUndefinedBehavior ub -> UB.ppReg ub
+  BBUndefinedBehavior ub -> UB.ppDetails ub
   BBMemoryError me -> ME.ppMemoryError me

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Errors/UndefinedBehavior.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Errors/UndefinedBehavior.hs
@@ -44,8 +44,11 @@ module Lang.Crucible.LLVM.Errors.UndefinedBehavior
     PtrComparisonOperator(..)
   , UndefinedBehavior(..)
   , cite
+  , details
   , explain
-  , ppReg
+  , ppDetails
+  , ppCitation
+  , pp
 
   , concUB
   ) where
@@ -370,10 +373,10 @@ explain =
 
 -- | Pretty-print the additional information held by the constructors
 -- (for symbolic expressions)
-detailsReg :: W4I.IsExpr (W4I.SymExpr sym)
-           => UndefinedBehavior (RegValue' sym)
-           -> [Doc]
-detailsReg =
+details :: W4I.IsExpr (W4I.SymExpr sym)
+        => UndefinedBehavior (RegValue' sym)
+        -> [Doc]
+details =
   \case
 
     -------------------------------- Memory management
@@ -463,23 +466,24 @@ detailsReg =
 pp :: (UndefinedBehavior e -> [Doc]) -- ^ Printer for constructor data
    -> UndefinedBehavior e
    -> Doc
-pp extra ub = vcat $
-    explain ub
-  : extra ub
-  ++ cat [ "Reference: "
-         , indent 2 (text (unpack (ppStd (standard ub))))
-         , indent 2 (cite ub)
-         ]
-     : case stdURL (standard ub) of
-         Just url -> [ indent 2 ("Document URL:" <+> text (unpack url)) ]
-         Nothing  -> []
+pp extra ub = vcat (explain ub : extra ub ++ ppCitation ub)
 
 -- | Pretty-printer for symbolic backends
-ppReg ::
+ppDetails ::
   W4I.IsExpr (W4I.SymExpr sym) =>
   UndefinedBehavior (RegValue' sym) ->
   Doc
-ppReg = pp detailsReg
+ppDetails ub = vcat (details ub ++ ppCitation ub)
+
+ppCitation :: UndefinedBehavior e -> [Doc]
+ppCitation ub =
+   (cat [ "Reference: "
+        , indent 2 (text (unpack (ppStd (standard ub))))
+        , indent 2 (cite ub)
+        ]
+    : case stdURL (standard ub) of
+        Just url -> [ indent 2 ("Document URL:" <+> text (unpack url)) ]
+        Nothing  -> [])
 
 -- -----------------------------------------------------------------------
 -- ** Instances

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Eval.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Eval.hs
@@ -19,6 +19,7 @@ import           Lang.Crucible.Simulator.RegValue
 import           Lang.Crucible.Simulator.SimError
 
 import qualified Lang.Crucible.LLVM.Arch.X86 as X86
+import qualified Lang.Crucible.LLVM.Errors.UndefinedBehavior as UB
 import           Lang.Crucible.LLVM.Extension
 import           Lang.Crucible.LLVM.MemModel.Pointer
 import           Lang.Crucible.LLVM.MemModel.Partial
@@ -30,7 +31,7 @@ assertSideCondition ::
   IO ()
 assertSideCondition sym (LLVMSideCondition (RV p) ub) =
   do p' <- annotateUB sym ub p
-     let err = AssertFailureSimError "Undefined behavior encountered" ""
+     let err = AssertFailureSimError "Undefined behavior encountered" (show (UB.explain ub))
      assert sym p' err
 
 llvmExtensionEval :: forall sym arch.

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
@@ -222,6 +222,9 @@ declare_overrides =
   , polymorphic1_llvm_override "llvm.umul.with.overflow"
       (\w -> SomeLLVMOverride (LLVM.llvmUmulWithOverflow w))
 
+  , basic_llvm_override LLVM.llvmFabsF32
+  , basic_llvm_override LLVM.llvmFabsF64
+
   -- C standard library functions
   , basic_llvm_override Libc.llvmAbortOverride
   , basic_llvm_override Libc.llvmAssertRtnOverride

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
@@ -39,6 +39,7 @@ import qualified Data.Parameterized.Context as Ctx
 import           Data.Parameterized.Context ( pattern (:>), pattern Empty )
 
 import           What4.Interface
+import           What4.InterpretedFloatingPoint
 
 import           Lang.Crucible.Backend
 import           Lang.Crucible.CFG.Common (GlobalVar)
@@ -516,6 +517,29 @@ llvmBSwapOverride widthRepr =
                 vec = regValue (args^._1)
             in bvSwap sym widthRepr vec)
     }}}
+
+
+
+llvmFabsF32
+  :: forall sym wptr arch p
+   . ( IsSymInterface sym, HasPtrWidth wptr, wptr ~ ArchWidth arch)
+  => LLVMOverride p sym arch
+        (EmptyCtx ::> FloatType SingleFloat)
+        (FloatType SingleFloat)
+llvmFabsF32 =
+  [llvmOvr| float @llvm.fabs.f32( float ) |]
+  (\_memOps sym (Empty :> (regValue -> x)) -> liftIO (iFloatAbs @_ @SingleFloat sym x))
+
+
+llvmFabsF64
+  :: forall sym wptr arch p
+   . ( IsSymInterface sym, HasPtrWidth wptr, wptr ~ ArchWidth arch)
+  => LLVMOverride p sym arch
+        (EmptyCtx ::> FloatType DoubleFloat)
+        (FloatType DoubleFloat)
+llvmFabsF64 =
+  [llvmOvr| double @llvm.fabs.f64( double ) |]
+  (\_memOps sym (Empty :> (regValue -> x)) -> liftIO (iFloatAbs @_ @DoubleFloat sym x))
 
 
 llvmX86_pclmulqdq

--- a/crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs
@@ -229,7 +229,8 @@ import           Lang.Crucible.Simulator.SimError
 import           Lang.Crucible.LLVM.DataLayout
 import           Lang.Crucible.LLVM.Extension
 import           Lang.Crucible.LLVM.Bytes
-import           Lang.Crucible.LLVM.Errors.MemoryError (MemErrContext, MemoryErrorReason(..), MemoryOp(..))
+import           Lang.Crucible.LLVM.Errors.MemoryError
+   (MemErrContext, MemoryErrorReason(..), MemoryOp(..), ppMemoryErrorReason)
 import qualified Lang.Crucible.LLVM.Errors.UndefinedBehavior as UB
 import           Lang.Crucible.LLVM.MemType
 import qualified Lang.Crucible.LLVM.MemModel.MemLog as ML
@@ -303,7 +304,7 @@ assertUndefined ::
   IO ()
 assertUndefined sym p ub =
   do p' <- Partial.annotateUB sym ub p
-     assert sym p' $ AssertFailureSimError "Undefined behavior encountered" ""
+     assert sym p' $ AssertFailureSimError "Undefined behavior encountered" (show (UB.explain ub))
 
 
 assertStoreError ::
@@ -315,7 +316,7 @@ assertStoreError ::
   IO ()
 assertStoreError sym errCtx rsn p =
   do p' <- Partial.annotateME sym errCtx rsn p
-     assert sym p' $ AssertFailureSimError "Memory store failed" ""
+     assert sym p' $ AssertFailureSimError "Memory store failed" (show (ppMemoryErrorReason rsn))
 
 instance IsSymInterface sym => IntrinsicClass sym "LLVM_memory" where
   type Intrinsic sym "LLVM_memory" ctx = MemImpl sym
@@ -437,7 +438,7 @@ evalStmt sym = eval
            Left doc -> lift $
              do p <- Partial.annotateME sym mop (BadFunctionPointer doc) (falsePred sym)
                 loc <- getCurrentProgramLoc sym
-                let err = SimError loc (AssertFailureSimError "Failed to load function handle" "")
+                let err = SimError loc (AssertFailureSimError "Failed to load function handle" (show doc))
                 addProofObligation sym (LabeledPred p err)
                 abortExecBecause $ AssumedFalse $ AssumingNoError err
 

--- a/crucible/src/Lang/Crucible/Simulator/SimError.hs
+++ b/crucible/src/Lang/Crucible/Simulator/SimError.hs
@@ -20,6 +20,7 @@ module Lang.Crucible.Simulator.SimError (
     SimErrorReason(..)
   , SimError(..)
   , simErrorReasonMsg
+  , simErrorDetailsMsg
   , ppSimError
   ) where
 

--- a/crux-llvm/c-src/print-model.c
+++ b/crux-llvm/c-src/print-model.c
@@ -5,24 +5,22 @@
 int main () {
   size_t i;
   for (i = 0; i < crux_value_num(int8_t); ++i)
-    printf("%s = %"PRId8"\n", crux_names(int8_t)[i], crux_values(int8_t)[i]);
+    printf("%s = %"PRId8", %"PRIu8", 0x%"PRIx8"\n", crux_names(int8_t)[i], crux_values(int8_t)[i], crux_values(int8_t)[i], crux_values(int8_t)[i]);
 
   for (i = 0; i < crux_value_num(int16_t); ++i)
-    printf("%s = %"PRId16"\n", crux_names(int16_t)[i], crux_values(int16_t)[i]);
+    printf("%s = %"PRId16", %"PRIu16", 0x%"PRIx16"\n", crux_names(int16_t)[i], crux_values(int16_t)[i], crux_values(int16_t)[i], crux_values(int16_t)[i]);
 
   for (i = 0; i < crux_value_num(int32_t); ++i)
-    printf("%s = %"PRId32"\n", crux_names(int32_t)[i], crux_values(int32_t)[i]);
+    printf("%s = %"PRId32", %"PRIu32", 0x%"PRIx32"\n", crux_names(int32_t)[i], crux_values(int32_t)[i], crux_values(int32_t)[i], crux_values(int32_t)[i]);
 
   for (i = 0; i < crux_value_num(int64_t); ++i)
-    printf("%s = %"PRId64"\n", crux_names(int64_t)[i], crux_values(int64_t)[i]);
+    printf("%s = %"PRId64", %"PRIu64", 0x%"PRIx64"\n", crux_names(int64_t)[i], crux_values(int64_t)[i], crux_values(int64_t)[i], crux_values(int64_t)[i]);
 
   for (i = 0; i < crux_value_num(float); ++i)
-    printf("%s = %f\n", crux_names(float)[i], crux_values(float)[i]);
+    printf("%s = %f, %a\n", crux_names(float)[i], crux_values(float)[i], crux_values(float)[i]);
 
   for (i = 0; i < crux_value_num(double); ++i)
-    printf("%s = %f\n", crux_names(double)[i], crux_values(double)[i]);
+    printf("%s = %f, %a\n", crux_names(double)[i], crux_values(double)[i], crux_values(double)[i]);
 
   return 0;
 }
-
-

--- a/crux-llvm/src/Crux/LLVM/Compile.hs
+++ b/crux-llvm/src/Crux/LLVM/Compile.hs
@@ -1,4 +1,7 @@
+{-# LANGUAGE DataKinds #-}
 {-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
 module Crux.LLVM.Compile where
 
 import Control.Exception
@@ -6,8 +9,10 @@ import Control.Exception
 import Control.Monad
   ( unless, forM_ )
 import qualified Data.Binary.IEEE754 as IEEE754
+import           Data.BitVector.Sized (BV)
 import qualified Data.BitVector.Sized as BV
 import qualified Data.Foldable as Fold
+import qualified Numeric as N
 import Data.List
   ( intercalate, isSuffixOf )
 import qualified Data.Parameterized.Map as MapF
@@ -189,18 +194,54 @@ buildModelExes cruxOpts llvmOpts suff counter_src =
 
      return (printExe, debugExe)
 
+showBVLiteral :: (1 <= w) => NatRepr w -> BV w -> String
+showBVLiteral w bv =
+    (if x < 0 then "-0x" else "0x") ++ N.showHex i (if natValue w == 64 then "L" else "")
+
+  where
+  x = BV.asSigned w bv
+  i = abs x
+
+showFloatLiteral :: BV 32 -> String
+showFloatLiteral bv
+   | isNaN x          = "NAN"
+   | isInfinite x     = if signum x < 0 then "-INFINITY" else "INFINITY"
+   | isNegativeZero x = "-0.0f"
+   | otherwise        =
+      (if mag < 0 then "-0x" else "0x") ++ N.showHex (abs mag) ("p" ++ show ex ++ "f")
+
+ where
+ x = IEEE754.wordToFloat (fromInteger (BV.asUnsigned bv))
+ (mag,ex) = decodeFloat x
+
+
+showDoubleLiteral :: BV 64 -> String
+showDoubleLiteral bv
+   | isNaN x          = "(double) NAN"
+   | isInfinite x     = if signum x < 0 then "- ((double) INFINITY)" else "(double) INFINITY"
+   | isNegativeZero x = "-0.0"
+   | otherwise        =
+      (if mag < 0 then "-0x" else "0x") ++ N.showHex mag ("p" ++ show ex)
+
+ where
+ x = IEEE754.wordToDouble (fromInteger (BV.asUnsigned bv))
+ (mag,ex) = decodeFloat x
+
 ppValsC :: BaseTypeRepr ty -> Vals ty -> String
 ppValsC ty (Vals xs) =
   let (cty, cnm, ppRawVal) = case ty of
         BaseBVRepr n ->
-          ("int" ++ show n ++ "_t", "int" ++ show n ++ "_t", show)
+          ("int" ++ show n ++ "_t", "int" ++ show n ++ "_t", showBVLiteral n)
         BaseFloatRepr (FloatingPointPrecisionRepr eb sb)
-          | natValue eb == 8, natValue sb == 24
-          -> ("float", "float", show . IEEE754.wordToFloat . fromInteger . BV.asUnsigned)
+          | Just Refl <- testEquality eb (knownNat @8)
+          , Just Refl <- testEquality sb (knownNat @24)
+          -> ("float", "float", showFloatLiteral)
         BaseFloatRepr (FloatingPointPrecisionRepr eb sb)
-          | natValue eb == 11, natValue sb == 53
-          -> ("double", "double", show . IEEE754.wordToDouble . fromInteger . BV.asUnsigned)
+          | Just Refl <- testEquality eb (knownNat @11)
+          , Just Refl <- testEquality sb (knownNat @53)
+          -> ("double", "double", showDoubleLiteral)
         BaseRealRepr -> ("double", "real", (show . toDouble))
+
         _ -> error ("Type not implemented: " ++ show ty)
   in unlines
       [ "size_t const crucible_values_number_" ++ cnm ++
@@ -217,6 +258,7 @@ ppModelC :: ModelView -> String
 ppModelC m = unlines
              $ "#include <stdint.h>"
              : "#include <stddef.h>"
+             : "#include <math.h>"
              : ""
              : MapF.foldrWithKey (\k v rest -> ppValsC k v : rest) [] vals
             where vals = modelVals m

--- a/crux-llvm/src/Crux/LLVM/Compile.hs
+++ b/crux-llvm/src/Crux/LLVM/Compile.hs
@@ -12,7 +12,6 @@ import qualified Data.Binary.IEEE754 as IEEE754
 import           Data.BitVector.Sized (BV)
 import qualified Data.BitVector.Sized as BV
 import qualified Data.Foldable as Fold
-import qualified Numeric as N
 import Data.List
   ( intercalate, isSuffixOf )
 import qualified Data.Parameterized.Map as MapF
@@ -31,7 +30,7 @@ import What4.ProgramLoc
 import Lang.Crucible.Simulator.SimError
 
 import Crux
-import Crux.Model( toDouble )
+import Crux.Model( toDouble, showBVLiteral, showFloatLiteral, showDoubleLiteral )
 import Crux.Types
 
 import Crux.LLVM.Config
@@ -194,38 +193,6 @@ buildModelExes cruxOpts llvmOpts suff counter_src =
 
      return (printExe, debugExe)
 
-showBVLiteral :: (1 <= w) => NatRepr w -> BV w -> String
-showBVLiteral w bv =
-    (if x < 0 then "-0x" else "0x") ++ N.showHex i (if natValue w == 64 then "L" else "")
-
-  where
-  x = BV.asSigned w bv
-  i = abs x
-
-showFloatLiteral :: BV 32 -> String
-showFloatLiteral bv
-   | isNaN x          = "NAN"
-   | isInfinite x     = if signum x < 0 then "-INFINITY" else "INFINITY"
-   | isNegativeZero x = "-0.0f"
-   | otherwise        =
-      (if mag < 0 then "-0x" else "0x") ++ N.showHex (abs mag) ("p" ++ show ex ++ "f")
-
- where
- x = IEEE754.wordToFloat (fromInteger (BV.asUnsigned bv))
- (mag,ex) = decodeFloat x
-
-
-showDoubleLiteral :: BV 64 -> String
-showDoubleLiteral bv
-   | isNaN x          = "(double) NAN"
-   | isInfinite x     = if signum x < 0 then "- ((double) INFINITY)" else "(double) INFINITY"
-   | isNegativeZero x = "-0.0"
-   | otherwise        =
-      (if mag < 0 then "-0x" else "0x") ++ N.showHex mag ("p" ++ show ex)
-
- where
- x = IEEE754.wordToDouble (fromInteger (BV.asUnsigned bv))
- (mag,ex) = decodeFloat x
 
 ppValsC :: BaseTypeRepr ty -> Vals ty -> String
 ppValsC ty (Vals xs) =

--- a/crux-llvm/src/Crux/LLVM/Compile.hs
+++ b/crux-llvm/src/Crux/LLVM/Compile.hs
@@ -8,9 +8,6 @@ import Control.Exception
   ( SomeException(..), try )
 import Control.Monad
   ( unless, forM_ )
-import qualified Data.Binary.IEEE754 as IEEE754
-import           Data.BitVector.Sized (BV)
-import qualified Data.BitVector.Sized as BV
 import qualified Data.Foldable as Fold
 import Data.List
   ( intercalate, isSuffixOf )

--- a/crux/src/Crux/Model.hs
+++ b/crux/src/Crux/Model.hs
@@ -1,11 +1,11 @@
 -- | This file is almost exactly the same as crucible-c/src/Model.hs
 
 {-# Language DataKinds #-}
-
+{-# Language PolyKinds #-}
 {-# Language Rank2Types #-}
 {-# Language TypeFamilies #-}
 {-# Language TypeApplications #-}
-{-# Language PolyKinds #-}
+{-# Language TypeOperators #-}
 {-# Language ScopedTypeVariables #-}
 
 {-# OPTIONS_GHC -fno-warn-orphans #-}
@@ -14,13 +14,15 @@ module Crux.Model where
 
 import Data.Binary.IEEE754 as IEEE754
 import qualified Data.BitVector.Sized as BV
+import           Data.BitVector.Sized (BV)
 import Data.Parameterized.NatRepr(knownNat,natValue)
 import Data.Parameterized.TraversableF(traverseF)
 import Data.Parameterized.Map (MapF)
 import Data.Parameterized.Pair(Pair(..))
 import qualified Data.Parameterized.Map as MapF
+import qualified Numeric as N
 
-import Lang.Crucible.Types(BaseTypeRepr(..),BaseToType,FloatPrecisionRepr(..))
+import Lang.Crucible.Types
 import Lang.Crucible.Simulator.RegMap(RegValue)
 import What4.Expr (GroundEvalFn(..), ExprBuilder)
 import What4.ProgramLoc
@@ -75,20 +77,56 @@ evalModel ev (Model mp) = traverseF (evalVars ev) mp
 toDouble :: Rational -> Double
 toDouble = fromRational
 
+
+showBVLiteral :: (1 <= w) => NatRepr w -> BV w -> String
+showBVLiteral w bv =
+    (if x < 0 then "-0x" else "0x") ++ N.showHex i (if natValue w == 64 then "L" else "")
+
+  where
+  x = BV.asSigned w bv
+  i = abs x
+
+showFloatLiteral :: BV 32 -> String
+showFloatLiteral bv
+   | isNaN x          = "NAN"
+   | isInfinite x     = if signum x < 0 then "-INFINITY" else "INFINITY"
+   | isNegativeZero x = "-0.0f"
+   | otherwise        =
+      (if mag < 0 then "-0x" else "0x") ++ N.showHex (abs mag) ("p" ++ show ex ++ "f")
+
+ where
+ x = IEEE754.wordToFloat (fromInteger (BV.asUnsigned bv))
+ (mag,ex) = decodeFloat x
+
+
+showDoubleLiteral :: BV 64 -> String
+showDoubleLiteral bv
+   | isNaN x          = "(double) NAN"
+   | isInfinite x     = if signum x < 0 then "- ((double) INFINITY)" else "(double) INFINITY"
+   | isNegativeZero x = "-0.0"
+   | otherwise        =
+      (if mag < 0 then "-0x" else "0x") ++ N.showHex mag ("p" ++ show ex)
+
+ where
+ x = IEEE754.wordToDouble (fromInteger (BV.asUnsigned bv))
+ (mag,ex) = decodeFloat x
+
+
 ppValsJS :: FilePath -> BaseTypeRepr ty -> Vals ty -> [String]
 ppValsJS cwd ty (Vals xs) =
   let showEnt = case ty of
-        BaseBVRepr n -> showEnt' show n
+        BaseBVRepr n -> showEnt' (showBVLiteral n) n
         BaseFloatRepr (FloatingPointPrecisionRepr eb sb)
-          | natValue eb == 8, natValue sb == 24 -> showEnt'
-            (show . IEEE754.wordToFloat . fromInteger . BV.asUnsigned)
-            (knownNat @32)
+          | Just Refl <- testEquality eb (knownNat @8)
+          , Just Refl <- testEquality sb (knownNat @24)
+          -> showEnt' showFloatLiteral (32 :: Int)
         BaseFloatRepr (FloatingPointPrecisionRepr eb sb)
-          | natValue eb == 11, natValue sb == 53 -> showEnt'
-            (show . IEEE754.wordToDouble . fromInteger . BV.asUnsigned)
-            (knownNat @64)
+          | Just Refl <- testEquality eb (knownNat @11)
+          , Just Refl <- testEquality sb (knownNat @53)
+          -> showEnt' showDoubleLiteral (64 :: Int)
         BaseRealRepr -> showEnt' (show . toDouble) (knownNat @64)
         _ -> error ("Type not implemented: " ++ show ty)
+
   in map showEnt xs
 
   where
@@ -100,7 +138,6 @@ ppValsJS cwd ty (Vals xs) =
       , "val"  ~> jsStr (repr (entryValue e))
       , "bits" ~> jsStr (show n)
       ]
-
 
 ppModelJS :: FilePath -> ModelView -> String
 ppModelJS cwd m = case ents of

--- a/crux/src/Crux/Model.hs
+++ b/crux/src/Crux/Model.hs
@@ -112,8 +112,8 @@ showDoubleLiteral bv
  (mag,ex) = decodeFloat x
 
 
-ppValsJS :: FilePath -> BaseTypeRepr ty -> Vals ty -> [String]
-ppValsJS cwd ty (Vals xs) =
+valsJS :: FilePath -> BaseTypeRepr ty -> Vals ty -> IO [JS]
+valsJS cwd ty (Vals xs) =
   let showEnt = case ty of
         BaseBVRepr n -> showEnt' (showBVLiteral n) n
         BaseFloatRepr (FloatingPointPrecisionRepr eb sb)
@@ -127,27 +127,22 @@ ppValsJS cwd ty (Vals xs) =
         BaseRealRepr -> showEnt' (show . toDouble) (knownNat @64)
         _ -> error ("Type not implemented: " ++ show ty)
 
-  in map showEnt xs
+  in mapM showEnt xs
 
   where
-  showEnt' :: Show b => (a -> String) -> b -> Entry a -> String
+  showEnt' :: Show b => (a -> String) -> b -> Entry a -> IO JS
   showEnt' repr n e =
-    renderJS $ jsObj
-      [ "name" ~> jsStr (entryName e)
-      , "loc"  ~> jsLoc cwd (entryLoc e)
-      , "val"  ~> jsStr (repr (entryValue e))
-      , "bits" ~> jsStr (show n)
-      ]
+    do l <- jsLoc cwd (entryLoc e)
+       pure $ jsObj
+         [ "name" ~> jsStr (entryName e)
+         , "loc"  ~> l
+         , "val"  ~> jsStr (repr (entryValue e))
+         , "bits" ~> jsStr (show n)
+         ]
 
-ppModelJS :: FilePath -> ModelView -> String
-ppModelJS cwd m = case ents of
-                [] -> "[]"
-                _  -> unlines $ zipWith (++) pre ents ++ ["]"]
-  where vals = modelVals m
-        ents = MapF.foldrWithKey (\k v rest -> ppValsJS cwd k v ++ rest) [] vals
-        pre  = "[ " : repeat ", "
-
-
+modelJS :: FilePath -> ModelView -> IO JS
+modelJS cwd m =
+  jsList . concat <$> sequence (MapF.foldrWithKey (\k v xs -> valsJS cwd k v : xs) [] (modelVals m))
 
 instance Semigroup (Model sym) where
   (Model m1) <> m2        = MapF.foldrWithKey f m2 m1 where

--- a/crux/src/Crux/Report.hs
+++ b/crux/src/Crux/Report.hs
@@ -4,7 +4,7 @@
 module Crux.Report where
 
 import System.FilePath
-import System.Directory (createDirectoryIfMissing, getCurrentDirectory, makeAbsolute)
+import System.Directory (createDirectoryIfMissing, getCurrentDirectory, canonicalizePath)
 import System.IO
 import qualified Data.Foldable as Fold
 import Data.List (partition)
@@ -20,7 +20,7 @@ import Lang.Crucible.Backend
 import Crux.Types
 import Crux.Config.Common
 import Crux.Loops
-import Crux.Model ( ppModelJS )
+import Crux.Model ( modelJS )
 
 -- Note these should be data files. However, cabal-new build doesn't make it easy for the installation
 -- to find data files, so they are embedded as Text constants instead.
@@ -38,8 +38,8 @@ generateReport opts res
        createDirectoryIfMissing True (outDir opts)
        maybeGenerateSource opts (inputFiles opts)
        cwd <- getCurrentDirectory
-       writeFile (outDir opts </> "report.js")
-          $ "var goals = " ++ renderJS (jsList (renderSideConds opts cwd xs))
+       scs <- renderSideConds opts cwd xs
+       writeFile (outDir opts </> "report.js") $ "var goals = " ++ renderJS (jsList scs)
        T.writeFile (outDir opts </> "index.html") indexHtml
        T.writeFile (outDir opts </> "jquery.min.js") jquery
 
@@ -56,7 +56,7 @@ maybeGenerateSource opts files =
      h <- openFile (outDir opts </> "source.js") WriteMode
      hPutStrLn h "var sources = ["
      forM_ renderFiles $ \file -> do
-       absFile <- makeAbsolute file
+       absFile <- canonicalizePath file
        txt <- readFile absFile
        hPutStr h $ "{\"label\":" ++ show (takeFileName absFile) ++ ","
        hPutStr h $ "\"name\":" ++ show absFile ++ ","
@@ -67,9 +67,11 @@ maybeGenerateSource opts files =
   `catch` \(SomeException {}) -> return ()
 
 
-renderSideConds :: CruxOptions -> FilePath -> Seq (ProcessedGoals, ProvedGoals b) -> [ JS ]
-renderSideConds opts cwd = concatMap (go [] . snd) . Fold.toList
+renderSideConds :: CruxOptions -> FilePath -> Seq (ProcessedGoals, ProvedGoals b) -> IO [ JS ]
+renderSideConds opts cwd seqGls = concatMapM (go [] . snd) (Fold.toList seqGls)
   where
+  concatMapM f xs = concat <$> mapM f xs
+
   flatBranch (Branch x y : more) = flatBranch (x : y : more)
   flatBranch (x : more)          = x : flatBranch more
   flatBranch []                  = []
@@ -80,22 +82,26 @@ renderSideConds opts cwd = concatMap (go [] . snd) . Fold.toList
 
   go path gs =
     case gs of
-      AtLoc pl _ gs1  -> go ((jsLoc cwd pl, pl) : path) gs1
+      AtLoc pl _ gs1  ->
+        do pl' <- jsLoc cwd pl
+           go ((pl', pl) : path) gs1
+
       Branch g1 g2 ->
-        let (now,rest) = partition isGoal (flatBranch [g1,g2])
-        in concatMap (go path) now ++ concatMap (go path) rest
+        let (now,rest) = partition isGoal (flatBranch [g1,g2]) in
+          (++) <$> concatMapM (go path) now <*> concatMapM (go path) rest
 
       Goal asmps conc triv proved
         | skipSuccessReports opts
         , Proved _ <- proved
-        -> []
+        -> pure []
 
         | skipIncompleteReports opts
         , (SimError _ (ResourceExhausted _), _) <- conc
-        -> []
+        -> pure []
 
         | otherwise
-        -> [ jsSideCond cwd apath asmps conc triv proved ]
+        -> do s <- jsSideCond cwd apath asmps conc triv proved
+              pure [s]
 
           where
             (ls,ps) = unzip (reverse path)
@@ -112,17 +118,22 @@ jsSideCond ::
   (SimError,String) ->
   Bool ->
   ProofResult b ->
-  JS
+  IO JS
 jsSideCond cwd path asmps (conc,_) triv status =
-  jsObj
-  [ "status"          ~> proved
-  , "counter-example" ~> example
-  , "goal"            ~> jsStr goalReason
-  , "location"        ~> jsLoc cwd (simErrorLoc conc)
-  , "assumptions"     ~> jsList (map mkAsmp asmps)
-  , "trivial"         ~> jsBool triv
-  , "path"            ~> jsList path
-  ]
+  do loc <- jsLoc cwd (simErrorLoc conc)
+     asmps' <- mapM mkAsmp asmps
+     ex <- example
+     pure $ jsObj
+       [ "status"          ~> proved
+       , "counter-example" ~> ex
+       , "goal"            ~> goalReason
+       , "details-short"   ~> goalDetails
+       , "location"        ~> loc
+       , "assumptions"     ~> jsList asmps'
+       , "trivial"         ~> jsBool triv
+       , "path"            ~> jsList path
+       , "details-long"    ~> longDetails
+       ]
   where
   proved = case (status, simErrorReason conc) of
              (Proved{}, _) -> jsStr "ok"
@@ -131,13 +142,23 @@ jsSideCond cwd path asmps (conc,_) triv status =
              (NotProved _ Nothing, _) -> jsStr "unknown"
 
   example = case status of
-             NotProved _ex (Just m) -> -- TODO! do something with this explanation
-               JS (ppModelJS cwd m)
-             _                  -> jsNull
+             NotProved _ex (Just m) -> modelJS cwd m
+             _                      -> pure jsNull
+
+  longDetails =
+     case status of
+       NotProved ex _ -> jsStr (show ex)
+       _ -> jsNull
 
   mkAsmp (asmp,_) =
-    jsObj [ "loc" ~> jsLoc cwd (assumptionLoc asmp)
-          -- , "text" ~> jsStr (show (ppAssumptionReason asmp))
-          ]
+    do l <- jsLoc cwd (assumptionLoc asmp)
+       pure $ jsObj
+         [ "loc" ~> l
+         , "text" ~> jsStr (show (ppAssumptionReason asmp))
+         ]
 
-  goalReason = simErrorReasonMsg (simErrorReason conc)
+  goalReason  = jsStr (simErrorReasonMsg (simErrorReason conc))
+  goalDetails
+     | null msg  = jsNull
+     | otherwise = jsStr msg
+    where msg = simErrorDetailsMsg (simErrorReason conc)

--- a/crux/src/Crux/Report.hs
+++ b/crux/src/Crux/Report.hs
@@ -39,7 +39,12 @@ generateReport opts res
        maybeGenerateSource opts (inputFiles opts)
        cwd <- getCurrentDirectory
        scs <- renderSideConds opts cwd xs
-       writeFile (outDir opts </> "report.js") $ "var goals = " ++ renderJS (jsList scs)
+       let contents = renderJS (jsList scs)
+       -- Due to CORS restrictions, the only current way of statically loading local data
+       -- is by including a <script> with the contents we want.
+       writeFile (outDir opts </> "report.js") $ "var goals = " ++ contents
+       -- However, for some purposes, having a JSON file is more practical.
+       writeFile (outDir opts </> "report.json") contents
        T.writeFile (outDir opts </> "index.html") indexHtml
        T.writeFile (outDir opts </> "jquery.min.js") jquery
 

--- a/crux/src/Crux/UI/IndexHtml.hs
+++ b/crux/src/Crux/UI/IndexHtml.hs
@@ -13,6 +13,7 @@ indexHtml = [r|
 <!DOCTYPE html>
 <html>
 <head>
+<meta charset="UTF-8">
 
 <script src="jquery.min.js"></script>
 <script src="source.js"></script>

--- a/crux/src/Crux/UI/IndexHtml.hs
+++ b/crux/src/Crux/UI/IndexHtml.hs
@@ -139,9 +139,19 @@ function drawGoals() {
     var li = $('<div/>')
             .addClass('clickable')
             .append( drawStatus(g.status)
-                   , $('<span/>').text(g.goal)
+                   , $('<abbr/>')
+                       .text(g.goal)
+                       .attr('title', g['details-short'])
                    )
     li.click(function() {
+      var errPane = $('#error-pane')
+      errPane.empty()
+      if( g['details-long'] ) {
+        errPane.append( $('<pre/>').text(g['details-long']) )
+      } else {
+        errPane.append( $('<pre/>').text(g['details-short']) )
+      }
+
       $('.highlight-assumed').removeClass('highlight-assumed')
       $('.selected').removeClass('selected')
       li.addClass('selected')
@@ -223,6 +233,13 @@ body { height: 100%; padding: 0; margin: 0; }
   height: 90%;
   overflow: auto;
   font-size: 16px;
+}
+
+#error-pane {
+  font-family: monospace;
+  font-size: 12px;
+  overflow: auto;
+  max-height: 20vh;
 }
 
 .file-btn {
@@ -327,6 +344,7 @@ ol.source-file>li:before {
 <body><div id="nav-bar"></div
 ><div id="right-pane"
   ><div id="file-pane"></div
+  ><div id="error-pane"></div
   ><div id="source-pane"></div
 ></div
 ></body>

--- a/crux/src/Crux/UI/JS.hs
+++ b/crux/src/Crux/UI/JS.hs
@@ -1,3 +1,4 @@
+{-# Language MultiWayIf #-}
 {-# Language OverloadedStrings #-}
 -- | Utilites for generating JSON
 module Crux.UI.JS where
@@ -5,23 +6,25 @@ module Crux.UI.JS where
 import Data.Text(unpack)
 import Data.List(intercalate)
 import Data.Maybe(fromMaybe)
-import System.FilePath(isRelative,(</>))
+import System.Directory( canonicalizePath )
+import System.FilePath( isRelative )
 
 import What4.ProgramLoc
 
-
-jsLoc :: FilePath -> ProgramLoc -> JS
+jsLoc :: FilePath -> ProgramLoc -> IO JS
 jsLoc cwd x =
   case plSourceLoc x of
-    SourcePos f l c -> jsObj [ "file" ~> jsStr fabsolute
-                             , "line" ~> jsStr (show l)
-                             , "col" ~> jsStr (show c)
-                             ]
-                       where fstr = unpack f
-                             fabsolute | null fstr = ""
-                                       | isRelative fstr = cwd </> fstr
-                                       | otherwise = fstr
-    _               -> jsNull
+    SourcePos f l c ->
+      do let fstr = unpack f
+         fabsolute <-
+            if | null fstr -> pure ""
+               | otherwise -> canonicalizePath fstr
+         pure $ jsObj
+           [ "file" ~> jsStr fabsolute
+           , "line" ~> jsStr (show l)
+           , "col"  ~> jsStr (show c)
+           ]
+    _ -> pure jsNull
 
 --------------------------------------------------------------------------------
 newtype JS = JS { renderJS :: String }

